File: lib/core/analysis/risk_calc.dart
import '../settings/app_settings.dart';

class RiskCalcResult {
  final double notionalUsdt;
  final double qty;
  final double leverage;
  final double marginUsdt;
  final double slPct; // ?˜ìˆ˜ë£??¬í•¨ ?ì ˆ%
  final double tpPct; // ?˜ìˆ˜ë£??¬í•¨ ëª©í‘œ%
  final double slUsdt;
  final double tpUsdt;

  const RiskCalcResult({
    required this.notionalUsdt,
    required this.qty,
    required this.leverage,
    required this.marginUsdt,
    required this.slPct,
    required this.tpPct,
    required this.slUsdt,
    required this.tpUsdt,
  });
}

/// ?”ë©´ ?œê¸°??ê°„ë‹¨ ê³„ì‚°ê¸?/// - FuState??entry/stop/target/leverage/qtyë¥?ê·¸ë?ë¡??¬ìš©
/// - ?˜ìˆ˜ë£??•ë³µ)??AppSettings.feeRoundTripë¡?ë°˜ì˜
class RiskCalc {
  static RiskCalcResult compute({
    required double entry,
    required double stop,
    required double target,
    required double qty,
    required double leverage,
  }) {
    final feeRt = AppSettings.feeRoundTrip;
    final notional = (qty * entry).abs();
    final margin = leverage <= 0 ? 0.0 : (notional / leverage);

    double pctLoss() {
      if (entry <= 0) return 0;
      final raw = ((entry - stop).abs() / entry);
      return ((raw + feeRt) * 100.0);
    }

    double pctGain() {
      if (entry <= 0) return 0;
      final raw = ((target - entry).abs() / entry);
      final net = (raw - feeRt);
      return (net < 0 ? 0 : net * 100.0);
    }

    final slPct = pctLoss();
    final tpPct = pctGain();
    final slUsdt = notional * (slPct / 100.0);
    final tpUsdt = notional * (tpPct / 100.0);

    return RiskCalcResult(
      notionalUsdt: notional,
      qty: qty,
      leverage: leverage,
      marginUsdt: margin,
      slPct: slPct,
      tpPct: tpPct,
      slUsdt: slUsdt,
      tpUsdt: tpUsdt,
    );
  }
}

File: lib/assets/riskCalc.ts
/**
 * Fulink Pro ULTRA RiskCalc 포팅.
 * entry/stop/target/qty/leverage → notional, margin, slPct, tpPct, slUsdt, tpUsdt.
 */
export type RiskCalcResult = {
  notionalUsdt: number;
  qty: number;
  leverage: number;
  marginUsdt: number;
  slPct: number;
  tpPct: number;
  slUsdt: number;
  tpUsdt: number;
};

const DEFAULT_FEE_ROUND_TRIP = 0.001; // 0.1% 왕복

export function computeRiskCalc(
  params: {
    entry: number;
    stop: number;
    target: number;
    qty: number;
    leverage: number;
  },
  feeRoundTrip = DEFAULT_FEE_ROUND_TRIP
): RiskCalcResult {
  const { entry, stop, target, qty, leverage } = params;
  const notional = Math.abs(qty * entry);
  const margin = leverage <= 0 ? 0 : notional / leverage;

  const rawSl = entry <= 0 ? 0 : Math.abs(entry - stop) / entry;
  const slPct = (rawSl + feeRoundTrip) * 100;

  const rawTp = entry <= 0 ? 0 : Math.abs(target - entry) / entry;
  const netTp = rawTp - feeRoundTrip;
  const tpPct = netTp < 0 ? 0 : netTp * 100;

  const slUsdt = notional * (slPct / 100);
  const tpUsdt = notional * (tpPct / 100);

  return {
    notionalUsdt: notional,
    qty,
    leverage,
    marginUsdt: margin,
    slPct,
    tpPct,
    slUsdt,
    tpUsdt,
  };
}
